drivers: ethernet: stm32: PTP L2 timestamping support

alsrgv opened the pull request:
Adds support for PTP timestamping to STM32 H7 and F7 families.  Depends on https://github.com/zephyrproject-rtos/hal_stm32/pull/116 bugfix for STM32 H7 functionality to work correctly.

Signed-off-by: Alex Sergeev <asergeev@carbonrobotics.com>

alsrgv committed:
drivers: ethernet: stm32: PTP L2 timestamping support

Add support for IEEE 1588-2008 (PTP L2) for STM32F7 and STM32H7
families.

Signed-off-by: Alex Sergeev <asergeev@carbonrobotics.com>

alsrgv committed:
doc: gptp: Add nucleo_f767zi and nucleo_h745zi_q as supported boards

Add nucleo_f767zi and nucleo_h745zi_q as supported boards in
documentation.

Signed-off-by: Alex Sergeev <asergeev@carbonrobotics.com>

alsrgv committed:
samples: net: gptp: Add nucleo_f767zi and nucleo_h745zi_q boards

Add nucleo_f767zi and nucleo_h745zi_q boards to gPTP samples.

Signed-off-by: Alex Sergeev <asergeev@carbonrobotics.com>

alsrgv commented:
That's a great news. I'd prefer to merge current implementation and once STM32 HAL-native version is ready, switch to that. We can help with testing - we use PTP pretty heavily.

jukkar reviewed (approved):
LGTM

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -43,7 +48,6 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #define PHY_BSR  ((uint16_t)0x0001U)  /*!< Transceiver Basic Status Register */
 #define PHY_LINKED_STATUS  ((uint16_t)0x0004U)  /*!< Valid link established */
 
-#define GET_FIRST_DMA_TX_DESC(heth)	(heth->Init.TxDesc)
```
I would prefer not to delete this macro. But it does not makes the functionality worse. Thus please let me know the reason to delete it.

alsrgv replied:
The difficulty with that macro is that for H7 it becomes all this code, which essentially has two outputs (cur_tx_desc_idx, dma_tx_desc):

```
	uint32_t cur_tx_desc_idx;

	cur_tx_desc_idx = heth->TxDescList.CurTxDesc;
	dma_tx_desc = (ETH_DMADescTypeDef *)heth->TxDescList.TxDesc[cur_tx_desc_idx];
```

Do you have a preference of how it should be handled instead?

Nukersson replied:
Ok. Fine. Let Dekrete it. The Code is anyway correct.

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -306,6 +354,69 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+		ETH_TxDescListTypeDef * dma_tx_desc_list;
+
+		__IO ETH_DMADescTypeDef *last_dma_tx_desc;
```
Are you sure __IO needed here? Usually it is STM32 HAL specific macro. Thus I would prefer to use C construct. As I remember __IO should be defined as volatile or so.

alsrgv replied:
I kept it around because eth_tx() and eth_rx() both have `__IO ETH_DMADescTypeDef` variables. It might be useful because this memory is populated concurrently with main CPU processing, so CPU shouldn't assume any caching guarantees about it?

Nukersson replied:
Well lets leave __IO

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -306,6 +354,69 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
+		__IO ETH_DMADescTypeDef *last_dma_tx_desc;
+
+		dma_tx_desc_list = &heth->TxDescList;
+		for (uint32_t i = 0; i < ETH_TX_DESC_CNT; i++) {
+			uint32_t last_desc_idx;
+
+			last_desc_idx = (cur_tx_desc_idx + i) % ETH_TX_DESC_CNT;
```
```suggestion
              uint32_t last_desc_idx;
		for (uint32_t i = 0; i < ETH_TX_DESC_CNT; i++) {
			last_desc_idx = (cur_tx_desc_idx + i) % ETH_TX_DESC_CNT;
```

... or use const uint32_t last_desc_idx = (cur_tx_desc_idx + i) % ETH_TX_DESC_CNT; in the cacle alternatively.

alsrgv replied:
Let me try const uint32_t option.

alsrgv replied:
I got dinged because of lack of space, so I had to add space after const uint32_t:
```
-:219: WARNING:LINE_SPACING: Missing a blank line after declarations
#219: FILE: drivers/ethernet/eth_stm32_hal.c:368:
+			const uint32_t last_desc_idx = (cur_tx_desc_idx + i) % ETH_TX_DESC_CNT;
+			last_dma_tx_desc =
```

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -306,6 +354,69 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
+			k_yield();
+		}
+
+		if (last_dma_tx_desc->DESC3 & ETH_DMATXNDESCWBF_LD &&
+				last_dma_tx_desc->DESC3 & ETH_DMATXNDESCWBF_TTSS) {
```
```suggestion
		if ((last_dma_tx_desc->DESC3 & ETH_DMATXNDESCWBF_LD) &&
		     (last_dma_tx_desc->DESC3 & ETH_DMATXNDESCWBF_TTSS)) {
```

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -306,6 +354,69 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
+			pkt->timestamp.nanosecond = UINT32_MAX;
+		}
+#else
+		__IO ETH_DMADescTypeDef *last_dma_tx_desc = dma_tx_desc;
```
... same with __IO as above.

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -388,6 +505,51 @@ static struct net_pkt *eth_rx(const struct device *dev, uint16_t *vlan_tag)
+	if (dma_rx_desc_list->AppDescNbr) {
+		uint32_t last_desc_idx;
+
+		__IO ETH_DMADescTypeDef *last_dma_rx_desc;
```
... __IO - see above.

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -388,6 +505,51 @@ static struct net_pkt *eth_rx(const struct device *dev, uint16_t *vlan_tag)
+
+	dma_rx_desc_list = &heth->RxDescList;
+	if (dma_rx_desc_list->AppDescNbr) {
+		uint32_t last_desc_idx;
+
+		__IO ETH_DMADescTypeDef *last_dma_rx_desc;
+
+		last_desc_idx =
+			(dma_rx_desc_list->FirstAppDesc + dma_rx_desc_list->AppDescNbr - 1U)
+				% ETH_RX_DESC_CNT;
```
```suggestion
		__IO ETH_DMADescTypeDef *last_dma_rx_desc;

		const uint32_t last_desc_idx =
			(dma_rx_desc_list->FirstAppDesc + dma_rx_desc_list->AppDescNbr - 1U)
				% ETH_RX_DESC_CNT;
```

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -388,6 +505,51 @@ static struct net_pkt *eth_rx(const struct device *dev, uint16_t *vlan_tag)
+		if (dma_rx_desc_list->AppContextDesc &&
+				last_dma_rx_desc->DESC1 & ETH_DMARXNDESCWBF_TSA) {
+			/* Retrieve timestamp from context DMA descriptor */
+			uint32_t context_desc_idx;
+
+			__IO ETH_DMADescTypeDef *context_dma_rx_desc;
+
+			context_desc_idx = (last_desc_idx + 1U) % ETH_RX_DESC_CNT;
```
```suggestion
			__IO ETH_DMADescTypeDef *context_dma_rx_desc;

			const uint32_t context_desc_idx = (last_desc_idx + 1U) % ETH_RX_DESC_CNT;
```

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -388,6 +505,51 @@ static struct net_pkt *eth_rx(const struct device *dev, uint16_t *vlan_tag)
+		}
+	}
+#else
+	__IO ETH_DMADescTypeDef *last_dma_rx_desc;
```
__IO ...

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -932,3 +1130,285 @@ static struct eth_stm32_hal_dev_data eth0_data = {
+		}
+		heth->Instance->MACTSCR |= ETH_MACTSCR_TSUPDT;
+		while (heth->Instance->MACTSCR & ETH_MACTSCR_TSUPDT_Msk) {
+			/* spin lock */
```
Depending on function's call place, it could be potentially dangerous. Maybe k_yield() needed here? Bu I am also not sure.

lochej replied:
Not sure if that's a great ideas since IRQs are locked in this context, would yielding back to scheduler without unlocking the irq actually be an issue ?

alsrgv replied:
I wanted to use irq_lock() here and particularly in `ptp_clock_stm32_get` because we use it in a zero-latency callback which doesn't allow any kernel code.  Basically, getting time should be as atomic operation as possible.

Nukersson replied:
Ok.

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -932,3 +1130,285 @@ static struct eth_stm32_hal_dev_data eth0_data = {
+		}
+		heth->Instance->PTPTSCR |= ETH_PTPTSCR_TSSTU;
+		while (heth->Instance->PTPTSCR & ETH_PTPTSCR_TSSTU_Msk) {
+			/* spin lock */
```
... same as above

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -932,3 +1130,285 @@ static struct eth_stm32_hal_dev_data eth0_data = {
+	ratio *= eth_dev_data->clk_ratio_adj;
+
+	/* Limit possible ratio */
+	if (ratio < 0.9 || ratio > 1.1) {
```
It would be good practice to put such values in Config with proper description.

alsrgv replied:
Updated using config. Had to use percentages instead of 0.9/1.1 because config doesn't seem to have floating point type support.

Nukersson replied:
Percentages are fine too.

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -932,3 +1130,285 @@ static struct eth_stm32_hal_dev_data eth0_data = {
+	heth->Instance->MACTSAR = addend_val;
+	heth->Instance->MACTSCR |= ETH_MACTSCR_TSADDREG;
+	while (heth->Instance->MACTSCR & ETH_MACTSCR_TSADDREG_Msk) {
+		/* spin lock */
```
... again - potentially dangerous waiting.

lochej replied:
Same comment about irq_lock ;)

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -932,3 +1130,285 @@ static struct eth_stm32_hal_dev_data eth0_data = {
+		&ptp_hclk_rate);
+	if (ret) {
+		LOG_ERR("Failed to query ethernet clock");
+		return -EIO;
```
Maybe ret itself should be returned here?

lochej replied:
Returning like this looks fine to me, we clearly know it's an IO error.

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -932,3 +1130,285 @@ static struct eth_stm32_hal_dev_data eth0_data = {
+
+	/* Program timestamp addend register */
+	eth_dev_data->clk_ratio =
+		(double)CONFIG_ETH_STM32_HAL_PTP_CLOCK_SRC_HZ / ptp_hclk_rate;
```
```suggestion
		((double)CONFIG_ETH_STM32_HAL_PTP_CLOCK_SRC_HZ) / ((double)ptp_hclk_rate);
```

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -932,3 +1130,285 @@ static struct eth_stm32_hal_dev_data eth0_data = {
+	/* Program timestamp addend register */
+	eth_dev_data->clk_ratio =
+		(double)CONFIG_ETH_STM32_HAL_PTP_CLOCK_SRC_HZ / ptp_hclk_rate;
+	eth_dev_data->clk_ratio_adj = 1.0;
```
Why is it 1.0?

alsrgv replied:
We determine `clk_ratio` as ratio between desired PTP frequency and HCLK frequency. Now, we might need to correct clock because of crystal misspec/drift - and that's what clk_ratio_adj is for. It starts as 1.0 and is allowed to be in bounds (0.9, 1.1).

Nukersson replied:
Please use define with meaningful name or put short comment Witz description here.

alsrgv replied:
Added a comment - please let me know if it doesn't explain it sufficiently.

KozhinovAlexander reviewed (changes requested):
Good work overall... I've not tried to compile and run it yet. I can do it in two weeks only. But if maybe @lochej already tested it, I would agree to merge.

alsrgv reviewed (commented):
Thanks for the feedback!  Here are updates based on first round, and few more questions in-line.

mbolivar-nordic reviewed (dismissed):
A soft -1 since I do not have high confidence in the devicetree usage -- I would appreciate @erwango taking a look and confirming it is OK.

Nukersson commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -112,4 +112,46 @@ config ETH_STM32_MODE_HALFDUPLEX
+	help
+	  Enable STM32 PTP clock support.
+
+if PTP_CLOCK_STM32_HAL
```
Please remove the if statement here and use depends on in appropriate config menus.

alsrgv replied:
Updated, thanks!

Nukersson commented:
@lochej do you have time to test it?

lochej commented:
> @lochej do you have time to test it?

I would love to, I'm short on time to tests them... I have the following boards to try. Which samples do you suggest for testing the new features ? I can test the ethernet basics with HTTP servers or others net samples though.
Nucleo-F767ZI
Nucleo-H723ZG
Nucleo-H743ZI

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -932,3 +1127,293 @@ static struct eth_stm32_hal_dev_data eth0_data = {
+	uint32_t addend_val;
+
+	/* No change needed */
+	if (ratio == 1.0) {
```
```suggestion
	if (ratio == 1.0f) {
```
Please update everywhere it is applicable

alsrgv replied:
Updated, thanks!

erwango reviewed (changes requested):
Nitpick, otherwise sounds globally good.

Nukersson commented:
@alsrgv What were your tests to check proper functionality of PTP in the driver?

alsrgv commented on doc/reference/networking/gptp.rst:
```diff
@@ -36,6 +36,8 @@ support must be enabled in ethernet drivers.
 Boards supported:
 
 - :ref:`frdm_k64f`
+- :ref:`nucleo_h745zi_q_board`
```
👍 added

alsrgv commented:
@Nukersson, I've tested it with gPTP sample, although I don't have fully compatible hardware with right accuracy so it was not very stable. If you have a setup where gPTP sample works with another board, e.g. FRDM, it should work on STM32 with this code as well.

We wrote IEEE 1588 PTPv2 slave-only driver that we've been testing this on (it requires both TX and RX timestamping to work correctly). We could push it as an example if it helps. We've been getting <100ns synchronization accuracy with that code with off-the-shelf non-PTP switches in the loop.

Nukersson commented:
> @Nukersson, I've tested it with gPTP sample, although I don't have fully compatible hardware with right accuracy so it was not very stable. If you have a setup where gPTP sample works with another board, e.g. FRDM, it should work on STM32 with this code as well.
> 
> We wrote IEEE 1588 PTPv2 slave-only driver that we've been testing this on (it requires both TX and RX timestamping to work correctly). We could push it as an example if it helps. We've been getting <100ns synchronization accuracy with that code with off-the-shelf non-PTP switches in the loop.

Thanks for fast response. An additional sample would be great.

lochej commented on samples/net/gptp/README.rst:
```diff
@@ -25,9 +25,10 @@ Building and Running
-embedded device like NXP FRDM-K64F or Atmel SAM-E70 Xplained. Note that gPTP is
-only supported for boards that have an Ethernet port and which has support for
-collecting timestamps for sent and received Ethernet frames.
+embedded device like NXP FRDM-K64F, Nucle-H743-ZI, Nucleo-H745ZI-Q, =
```
Minor typo:
```suggestion
embedded device like NXP FRDM-K64F, Nucleo-H743-ZI, Nucleo-H745ZI-Q, =
```

alsrgv replied:
Good catch!

erwango commented on samples/net/gptp/boards/nucleo_f767zi.conf:
```diff
@@ -0,0 +1,3 @@
+# STM32 driver settings
+CONFIG_ETH_STM32_HAL=y
```
This should not be needed, driver will be enabled based on `CONFIG_NETWORKING=y` and `&mac {status = "okay";};` combination.
Apply to all

alsrgv replied:
Good catch - just pushed that in a separate commit, please confirm that this is what you had in mind.

erwango replied:
That's fine, but please fix directly in the initial guilty commit.

erwango reviewed (changes requested):
Last minute request on my side

jukkar commented:
> We wrote IEEE 1588 PTPv2 slave-only driver that we've been testing this on (it requires both TX and RX timestamping to work correctly). We could push it as an example if it helps. We've been getting <100ns synchronization accuracy with that code with off-the-shelf non-PTP switches in the loop.

That would be great if you can contribute more samples for gPTP :+1:

Nukersson commented:
@erwango @jukkar have you tested the samples and the changes already on some of yours HW? If not please wait until tomorrow evening with the merge, I would like to test it on my h745 board.

alsrgv commented:
I've pushed our PTP slave code as an example in samples/net/ptp/slave. It requires an actual board to be tested (I verified on FRDM64 and STM F7/H7), and a PTPv2 master - which can be a Linux computer, preferably with NIC that supports hardware timestamping.

Note that H7 boards require this fix to avoid HAL race condition: https://github.com/zephyrproject-rtos/hal_stm32/pull/116

Is there a tool I can use to quickly reformat sample code to Zephyr code standards? We use clang-format internally with fairly long line limits.

erwango commented:
> Is there a tool I can use to quickly reformat sample code to Zephyr code standards? We use clang-format internally with fairly long line limits.

Nothing to do it for you, but compliance check can help in detecting what should be dine, cf https://github.com/zephyrproject-rtos/zephyr/runs/4304480548?check_suite_focus=true

> Note that H7 boards require this fix to avoid HAL race condition: zephyrproject-rtos/hal_stm32#116

Thanks for reminding, I'm trying to get this moving forward internally.

alsrgv commented:
Updated example to match project formatting.

erwango commented on samples/net/ptp/slave/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+project(ptp_slave)
+
+file(GLOB APP_SOURCES src/*.c)
+file(GLOB APP_HEADeRS src/*.h)
```
```suggestion
file(GLOB APP_HEADERS src/*.h)
```

jukkar commented on samples/net/ptp/slave/src/ptp_slave.h:
```diff
@@ -0,0 +1,84 @@
+#include <net/net_if.h>
+#include <net/ptp_time.h>
+
+#define PTP_MESSAGE_TYPE_SYNC 0x0
```
Various defines in this file are already found in `include/net/gptp.h` file, and if something is missing there then those can be added to that header file. I would rather see that the symbols in gptp.h are used here in the sample so we avoid defining same ones just for this sample.

jukkar commented on samples/net/ptp/slave/src/ptp_slave.c:
```diff
@@ -0,0 +1,313 @@
+
+	state.iface = iface;
+
+	state.clk = net_eth_get_ptp_clock(iface);
```
Please check that the returned value is not NULL, as it is possible that the interface does not support PTP (if there are multiple Ethernet interfaces in the system).

alsrgv replied:
Addressed in #41031

jukkar commented on samples/net/ptp/slave/src/ptp_slave.c:
```diff
@@ -0,0 +1,313 @@
+		return ret;
+	}
+
+	struct net_linkaddr *iface_addr = net_if_get_link_addr(state.iface);
```
Please introduce all variables at the beginning of the block.

jukkar commented on samples/net/ptp/slave/src/ptp_slave.c:
```diff
@@ -0,0 +1,313 @@
+	state.kp = 0.3;
+	state.kd = 0.7;
+
+	ret = net_context_get(AF_PACKET, SOCK_RAW, ETH_P_ALL, &state.context);
+	if (ret < 0) {
+		return ret;
+	}
+
+	struct sockaddr_ll dst;
+
+	dst.sll_ifindex = net_if_get_by_iface(iface);
+	dst.sll_family = AF_PACKET;
+	dst.sll_protocol = htons(NET_ETH_PTYPE_PTP);
+
+	ret = net_context_bind(state.context, (const struct sockaddr *)&dst, sizeof(dst));
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = net_context_recv(state.context, pkt_received, K_NO_WAIT, NULL);
+	if (ret < 0) {
+		return ret;
+	}
```
We try to avoid using `net_context` APIs in the samples as the BSD socket interface is meant for application use and can be used here.

alsrgv replied:
This code requires receiving all PTP L2 ethernet packets, does Zephyr's BSD socket API support that use case? Is there an example?

jukkar replied:
For raw socket support, see `samples/net/sockets/packet` for an example.

alsrgv replied:
Thanks for the pointer. I believe last time I tried this, PTP packets were confused with gPTP and dropped - so this might require some more modifications to ethernet stack. We can explore that as part of adding this example in a follow-on PR.

jukkar replied:
Ah, it might be that gPTP stack is stealing the packets but that should have happened when using `net_context` API too. Internally BSD socket code is just calling net_context API.

alsrgv replied:
I've started playing with that. Question - how to accomplish this with sockets API? `net_pkt_timestamp(pkt)`

jukkar reviewed (changes requested):
The sample application is good, it just needs some tweaking. One option to proceed is to split this PR and submit the sample app in the separate PR. This would speed up the merge of the driver part.

erwango commented:
@alsrgv Looks good to me now on the driver side. I'll let @jukkar and @rlubos review the sample part.
One last think I'd like is that  we can address the point of (functional) dependency with https://github.com/zephyrproject-rtos/hal_stm32/pull/116.
This will have some complication to this PR, but you look like a patient guy, looking to do things the right way, so let's tackle this last part:
What we need to do is to update zephyr repo to get it pointing to the hal update: https://github.com/zephyrproject-rtos/hal_stm32/pull/116. This will help building the hal change and people pulling this PR will also get the hal fix (using `west update`)
Please add a new commit updating `west.yml` file with the following change:
```
    - name: hal_stm32
-      revision: 6fc9b5f7e0229ef64a882e37213266d346ebfadb
+      revision: https://github.com/zephyrproject-rtos/hal_stm32/pull/116
      path: modules/hal/stm32
      groups:
        - hal
```
Once CI and tests confirm the hal fix is ok, we'll get it merged.

alsrgv commented:
@erwango, I tried to update west.yaml like this:
```
    - name: hal_stm32
      revision: https://github.com/zephyrproject-rtos/hal_stm32/pull/116
      path: modules/hal/stm32
      groups:
        - hal
```

However, I'm getting an error:
```
ERROR: update failed for project hal_stm32
```

Is there a better way to do it?

@jukkar, I like your idea of splitting the example part into the separate PR.  I think we should introduce `include/net/ptp.h` with proper definitions and make gPTP use some of those, since gPTP is a subset of PTP.

erwango commented:
> @erwango, I tried to update west.yaml like this

Opps my bad, change should rather be

 ```
     - name: hal_stm32
       revision: /pull/116/head
       path: modules/hal/stm32
       groups:
         - hal
 ```

alsrgv committed:
west: Update hal_stm32 to 5ca535df5874d1f9ab931265a2b1f691b76df9ce

PR hal_stm32 PR #116 avoids ethenet descriptor leak with enabled
timestamping.

Signed-off-by: Alex Sergeev <asergeev@carbonrobotics.com>

alsrgv commented:
@erwango, I think I found the way - updated west.yaml

alsrgv commented:
@jukkar @erwango, any updates on this?

erwango commented:
@alsrgv Thanks for the reminder, we'll try to complete review and get this in by this week

erwango commented:
@alsrgv https://github.com/zephyrproject-rtos/hal_stm32/pull/116 is now merged. Please update commit changing west.yml to make it target actual commit SHA1

alsrgv commented:
@erwango, done, thanks!

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -932,3 +1127,293 @@ static struct eth_stm32_hal_dev_data eth0_data = {
+		k_yield();
+	}
+#else
+	heth->Instance->PTPTSAR = addend_val;
+	heth->Instance->PTPTSCR |= ETH_PTPTSCR_TSARU;
+	while (heth->Instance->PTPTSCR & ETH_PTPTSCR_TSARU_Msk) {
+		k_yield();
+	}
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
+	/* Enable fine timestamp correction method */
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	heth->Instance->MACTSCR |= ETH_MACTSCR_TSCFUPDT;
+#else
+	heth->Instance->PTPTSCR |= ETH_PTPTSCR_TSFCU;
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
+	/* Enable nanosecond rollover into a new second */
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	heth->Instance->MACTSCR |= ETH_MACTSCR_TSCTRLSSR;
+#else
+	heth->Instance->PTPTSCR |= ETH_PTPTSSR_TSSSR;
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
+	/* Initialize timestamp */
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	heth->Instance->MACSTSUR = 0;
+	heth->Instance->MACSTNUR = 0;
+	heth->Instance->MACTSCR |= ETH_MACTSCR_TSINIT;
+	while (heth->Instance->MACTSCR & ETH_MACTSCR_TSINIT_Msk) {
+		k_yield();
+	}
+#else
+	heth->Instance->PTPTSHUR = 0;
+	heth->Instance->PTPTSLUR = 0;
+	heth->Instance->PTPTSCR |= ETH_PTPTSCR_TSSTI;
+	while (heth->Instance->PTPTSCR & ETH_PTPTSCR_TSSTI_Msk) {
+		k_yield();
+	}
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
```
Nit , I wonder if one big #if #else wouldn't be easier to read.

alsrgv replied:
I'd prefer to keep like this for now because steps across MCUs perfectly align, but registers don't. This makes it easier to make sure the right code is still executed. Can be revised if we find STM family that follows a different procedure that we want to support.

erwango reviewed (approved):
Fine on my side

jukkar commented:
@alsrgv could you send the commit `samples: net: ptp: slave: Add PTP slave sample` in a separate PR, the driver part looks ok but the slave sample requires a bit more work.

alsrgv commented:
@jukkar, done

cfriedt commented:
@alsrgv - looks like you may need to rebase / resolve conflicts.

There are also comments from @Nukersson and @lochej

alsrgv commented:
@cfriedt, rebased, thanks.

@Nukersson, @lochej - I believe I've addressed your comments. Is there anything outstanding that we should discuss before this is merged?

Nukersson commented:
> @cfriedt, rebased, thanks.
> 
> @Nukersson, @lochej - I believe I've addressed your comments. Is there anything outstanding that we should discuss before this is merged?

@alsrgv Thanks for well done contribution. I wanted test it also, but do not have time due to tight project schedule. If all the responsible persons agrees, it can be merged. By just skimming thought the code I did not found any wrong stuff.

cfriedt commented:
@alsrgv - one minor comment about init prio. Could it be changed to use an expression rather than hard-coded?

alsrgv commented:
@cfriedt, I took inspiration and default from MCUX's `ETH_MCUX_PTP_CLOCK_INIT_PRIO`. What should be the expression to use?

cfriedt commented:
> 



> @cfriedt, I took inspiration and default from MCUX's `ETH_MCUX_PTP_CLOCK_INIT_PRIO`. What should be the expression to use?

Would it be possible to use an expression like the one below?
```
default NET_INIT_PRIO - 1
```

alsrgv commented:
@cfriedt, I've dug around and I don't believe math expressions are supported as defaults in Kconfig - do you have an example?

cfriedt commented:
> @cfriedt, I've dug around and I don't believe math expressions are supported as defaults in Kconfig - do you have an example?

Yes, you're right sorry. In that case, maybe it's better to use

```
default NET_INIT_PRIO
```
and then subtract 1 in C

```
DEVICE_DEFINE(stm32_ptp_clock_0, PTP_CLOCK_NAME, ptp_stm32_init,
		NULL, &ptp_stm32_0_context, NULL, POST_KERNEL,
		CONFIG_ETH_STM32_HAL_PTP_CLOCK_INIT_PRIO - 1, &api);
```

alsrgv commented:
@cfriedt, sure, I can cut it out - I'll use `CONFIG_NET_INIT_PRIO - 5` to match existing value if that's alright.

alsrgv commented:
@cfriedt, I'm having trouble with that as well:
```
/opt/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: warning: orphan section `.z_device_POST_KERNEL90-5_' from `zephyr/drivers/ethernet/libdrivers__ethernet.a(eth_stm32_hal.c.obj)' being placed in section `.z_device_POST_KERNEL90-5_'
/opt/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: Undefined initialization levels used.
```

I feel like maybe that's the same process that mcux guys used before they just created another Kconfig setting for their PTP init priority.

cfriedt commented:
> @cfriedt, I'm having trouble with that as well:
> ```
> /opt/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: warning: orphan section `.z_device_POST_KERNEL90-5_' from `zephyr/drivers/ethernet/libdrivers__ethernet.a(eth_stm32_hal.c.obj)' being placed in section `.z_device_POST_KERNEL90-5_'
> /opt/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: Undefined initialization levels used.
> ```
> 
> I feel like maybe that's the same process that mcux guys used before they just created another Kconfig setting for their PTP init priority.

I think you're right. Sorry to ask you to go through the extra work

alsrgv commented on samples/net/ptp/slave/src/main.c:
```diff
@@ -0,0 +1,25 @@
+
+void main(void)
+{
+	struct net_if *iface = net_if_get_default();
```
Addressed in #41031
